Add CONTRIBUTING.md (#2833)
konflux-ci/build-definitions: CONTRIBUTING.md

## Project Overview

This repository serves as a central repository for several components used in [Konflux](https://konflux-ci.dev).
Therefore, everybody should follow its [Code of Conduct](https://github.com/konflux-ci/community/blob/main/CODE_OF_CONDUCT.md).

This central repository contains components managed by the Konflux Build Team and other various contributors.
See the CODEOWNERS file for more information.

Components are delivered to Konflux via the `konflux-ci/tekton-catalog` Quay organization as OCI bundles.

## How to Report Issues

We encourage early communication for all types of contributions.
If the contribution is non-trivial (straightforward bugfixes, typos, etc.), please open an issue to discuss your plans and get guidance from maintainers.

### Bug Reports

When reporting bugs, please include the following information:

**Required Information:**
- **Problem Description**: Clear, concise description of what went wrong
- **Steps to Reproduce**: Detailed steps to recreate the issue
- **Current Behavior**: What actually happened
- **Expected Behavior**: What you expected to happen
- **Other information**: For example relevant error messages or logs

> 

Reviewer: You're adding templates for both in the next commit, you sure this information needs to be duplicated instead of just saying please use one of our supplied templates to open an issue of a particular type?

### Feature Requests

For feature requests, please provide:

- **Use Case**: Describe the problem you're trying to solve
- **Proposed Solution**: Your suggested approach (if any)
- **Alternatives Considered**: Other solutions you've thought about
- **Impact**: Who would benefit from this feature

## How to Submit Pull Requests

### Development Workflow

1. **Fork and Clone**: Fork this repository and clone your fork

> 2. **Create Feature Branch**: Create a new branch from `main`

Reviewer: nitpick:
`s/new branch from main/new topic branch based on main`

3. **Make Changes**: Implement your changes
4. **Generate Content**: Run generation scripts if needed
   ```bash
   ./hack/generate-everything.sh
   ```
6. **Commit Changes**: See [commit guidelines](#pull-request-guidelines)

### Pull Request Guidelines

**Commit Requirements:**
- Write clear, descriptive commit titles. Should fit under 50 characters
- Write meaningful commit descriptions with each line having less than 72 characters
- Split your contribution into several commits if applicable, each should represent a logical chunk
- Add line `Assisted-by: <name-of-ai-tool>` if you used an AI tool for your contribution

**Pull Request Content:**
- **Title**: Clear, descriptive title. Should fit under 72 characters.

> - **Description**: Explain the changes and their purpose. You may refer to commit descriptions.

Reviewer: nitpick: I would suggest just a tiny adjustment to reduce the redundancy of information:
_Explain the overall changes and their purpose, this should be a cover letter for your commits_ ...

- **Testing**: Describe how the changes were tested
- **Links**: Reference related issues or upstream stories.

**Remember:**
- Konflux is a community project and proper descriptions cannot be replaced by referencing a publicly inaccessible link to Jira or any other private resource.
- Reviewers, other contributors and future generations might not have the same context as you have at the moment of PR submission.

### Code Quality Standards

**Shell Scripts:**
- Set `set -euo pipefail` at the beginning of bash scripts
- Pass shellcheck

> - Explain unintuitive code with comments

Reviewer: > I see your point, but haven't figured out more information I would like to include in something like HACKING.md and I don't feel like it's worthwhile to create a new file because of 5 lines.

Why so? Content amount doesn't really matter, the file can always grow as the consensus regarding the quality/guideline matures, so that's IMO irrelevant.

> Or I would also be fine with just removing the whole section, I don't think it's particularly useful.

Depends on how often you have to remind your contributors of this fact :).

  * [Pull Request Guidelines](#pull-request-guidelines)
  * [Code Quality Standards](#code-quality-standards)
  * [Security Best Practices](#security-best-practices)
* [Contribution Types](#contribution-types)
  * [Adding New Tasks](#adding-new-tasks)
  * [Modifying Existing Tasks](#modifying-existing-tasks)
  * [Adding New Pipelines](#adding-new-pipelines)
  * [Adding External Tasks](#adding-external-tasks)
  * [Task Deprecation](#task-deprecation)
* [Review Process](#review-process)

## Project Overview

This repository serves as a central repository for several components used in [Konflux](https://konflux-ci.dev).
Therefore, everybody should follow its [Code of Conduct](https://github.com/konflux-ci/community/blob/main/CODE_OF_CONDUCT.md).

This central repository contains components managed by the Konflux Build Team and other various contributors.
See the CODEOWNERS file for more information.

Components are delivered to Konflux via the `konflux-ci/tekton-catalog` Quay organization as OCI bundles.

Note that this repository will be undergoing two major changes:
1. Decentralization of tasks
2. Replacement of Build team's task scripts with a new Go CLI

## How to Report Issues

> We encourage early communication for all types of contributions. Before starting work, please open an issue to discuss your plans and get guidance from maintainers.

Reviewer 2: For straight-forward, small changes, submitting a PR directly is fine IMO. Maybe rephrase this so that it doesn't sound like we unconditionally want the contributor to open an issue first?

### Bug Reports

When reporting bugs, please include the following information:

> **Required Information:**

Reviewer 2: nitpick: the **Required Information** header seems unnecessary, since there is no optional information

- **Problem Description**: Clear, concise description of what went wrong
- **Steps to Reproduce**: Detailed steps to recreate the issue
- **Current Behavior**: What actually happened
- **Expected Behavior**: What you expected to happen
- **Other information**: For example relevant error messages or logs

**Template:**
```markdown
## Problem Description
Brief description of the issue

## Steps to Reproduce
1. Step one
2. Step two
3. Step three

## Current Behavior
What actually happens

## Expected Behavior
What should happen

## Other information
Logs, error messages, etc.

> ```

Reviewer 2: Github supports issue templates, so that you get something pre-filled when you open an issue. Would this template and the one below be more useful as actual issue templates?

[Github's docs on this](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/configuring-issue-templates-for-your-repository) are a bit weird and make it sound like issue templates are a click-to-set-up kind of affair, but I don't think that's true. ChatGPT makes it sound better: https://chatgpt.com/share/68da3eec-5900-800e-a36d-6103fdfd1c36

### Feature Requests

For feature requests, please provide:

**Required Information:**
- **Use Case**: Describe the problem you're trying to solve
- **Proposed Solution**: Your suggested approach (if any)
- **Alternatives Considered**: Other solutions you've thought about
- **Impact**: Who would benefit from this feature

**Template:**
```markdown
## Use Case
Describe the problem or need this feature would address

## Proposed Solution
Your suggested implementation approach

## Alternatives Considered
Other potential solutions

## Additional Context
Any other relevant information
```

## How to Submit Pull Requests

### Development Workflow

1. **Fork and Clone**: Fork this repository and clone your fork
2. **Create Feature Branch**: Create a new branch from `main`
3. **Make Changes**: Implement your changes following other guidelines

>     - Some tasks (like Buildah) may have several variants (buildah, buildah-oci-ta, buildah-remote...). Make changes only to the base version, e.g. `buildah.yaml`

Reviewer 2: What "other guidelines" are we referring to here? Just the list below?

4. **Generate Content**: Run generation scripts if needed
   ```bash
   ./hack/generate-everything.sh
   ```
6. **Commit Changes**: See [commit guidelines](#pull-request-guidelines)

### Pull Request Guidelines

**Commit Requirements:**
- Write clear, descriptive commit titles. Should fit under 50 characters
- Write meaningful commit descriptions with each line having less than 72 characters
- Split your contribution into several commits if applicable, each should represent a logical chunk

> - Add line `Assisted-by: <name-of-ai-tool> AI` if you used an AI tool for your contribution

Reviewer 2: ```suggestion
- Add line `Assisted-by: <name-of-ai-tool>` if you used an AI tool for your contribution
```

I don't think we care if the line has an `AI` suffix or not

**Pull Request Content:**

> - **Title**: Clear, descriptive title. Should fit under 50 characters.

Reviewer 2: The character limit for PR titles is not important IMO

- **Description**: Explain the changes and their purpose. You may refer to commit descriptions.
- **Testing**: Describe how the changes were tested
- **Links**: Reference related issues or upstream stories.

**Remember:**

> - Konflux is a community project and proper descriptions may not be replaced by referencing a publicly inaccessible link to Jira or any other private resource.

Reviewer 2: ```suggestion
- Konflux is a community project and proper descriptions cannot be replaced by referencing a publicly inaccessible link to Jira or any other private resource.
```
I think we can be more forceful here 😄

- Reviewers, other contributors and future generations might not have the same context as you have at the moment of PR submission.

### Code Quality Standards

**Shell Scripts:**
- Set `set -euo pipefail` at the beginning of bash scripts
- Pass shellcheck
- Explain unintuitive code with comments

### Security Best Practices

**Critical Requirements:**
- Never commit secrets or keys to the repository
- Never expose or log sensitive information

> ## Contribution Types

Reviewer 2: General feedback for this section: it duplicates some of the sections from README.md, but with an insufficient level of detail. I don't think an external contributor would know what to do based only on the instructions here.

I think it would make sense to move the necessary information from README.md to CONTRIBUTING.md. But IMO it's unnecessary to include this info at all for build-definitions, since many things will change after decentralization
